### PR TITLE
chat_with_ai changes

### PR DESCRIPTION
**Pull request by deadPix3l**
### Background
The chat_with_ai function was full of variables, create context returned a tuple of 4 variables(!!!) and the loop was just kind of hard to read. Additionally the try-except statement includes too much code

### Changes
- added reduce_dataset() to trim down memories and msg_history in a cleaner way
- moved much of the chat_with_ai code outside of the try-except as it doesnt reach out to openai, and cannot raise a rate limiting error. the try block now encompases ONLY the reachout.
- reworked generate_context to only return the context, not a bunch of other meta vars

### Documentation
inline comments. 

### Test Plan
ran before and after with same yaml files. ensured the loop didnt break and that prompts were acting on the new information from the last run

### PR Quality Checklist
- [X] My pull request is atomic and focuses on a single change.
- [X] I have thoroughly tested my changes with multiple different prompts.
- [X] I have considered potential risks and mitigations for my changes.
- [X] I have documented my changes clearly and comprehensively.
- [X] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: scripts/chat.py [hunk 1]
```diff
@@ -127,8 +126,12 @@ def chat_with_ai(
             )
 
             # Update full message history
-            full_message_history.append( create_chat_message( "user", user_input))
-            full_message_history.append( create_chat_message( "assistant", assistant_reply))
+            full_message_history.append(
```
Comment (deadPix3l) [inline scripts/chat.py]: undoing accidental "formatting-only" change introduced last commit

File: scripts/chat.py [hunk 2]
```diff
@@ -127,8 +126,12 @@ def chat_with_ai(
...
+                create_chat_message(
+                    "user", user_input))
+            full_message_history.append(
+                create_chat_message(
+                    "assistant", assistant_reply))
```
Comment (deadPix3l) [inline scripts/chat.py]: undoing accidental "formatting-only" change introduced last commit

File: scripts/chat.py [hunk 3]
```diff
@@ -53,72 +53,71 @@ def chat_with_ai(
...
+    user_input (str): The input from the user.
+    full_message_history (list): The list of all messages sent between the user and the AI.
+    permanent_memory (Obj): The memory object containing the permanent memory.
+    token_limit (int): The maximum number of tokens allowed in the API call.
+
+    Returns:
+    str: The AI's response.
+    """
     while True:
-        try:
```
Comment (deadPix3l) [inline scripts/chat.py:53]: This looks like a lot of changes. Its just indentation from moving the try block

**Comment by p-i-** [comment]
This is a mass message from the AutoGPT core team.
Our apologies for the ongoing delay in processing PRs.
This is because we are re-architecting the AutoGPT core!

For more details (and for infor on joining our Discord), please refer to:
  https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting

**Comment by Pwuts** [comment]
Thanks for the clean-up PR, it's appreciated! Since submission of this PR, this function has been refactored extensively, so these changes are hard to merge. I'll close this PR, but you're welcome to contribute again. If you would like to do so, be sure to get into contact via discord. :)

**Pull request closed without merging**